frame/election-provider-multi-phase/src/signed.rs

```diff
@@ -0,0 +1,707 @@
+			assert_eq!(balances(&99), (100, 0));
+
+			assert_ok!(submit_with_witness(Origin::signed(99), solution));
+			assert_eq!(balances(&99), (95, 5));
+
+			assert!(MultiPhase::finalize_signed_phase().0);
+			// expected reward is 5
+			assert_eq!(balances(&99), (100 + 5, 0));
+		});
+	}
+
+	#[test]
+	fn bad_solution_is_slashed() {
+		ExtBuilder::default().build_and_execute(|| {
+			roll_to(15);
+			assert!(MultiPhase::current_phase().is_signed());
+
+			let mut solution = raw_solution();
+			assert_eq!(balances(&99), (100, 0));
+
+			// make the solution invalid.
+			solution.score[0] += 1;
+
+			assert_ok!(submit_with_witness(Origin::signed(99), solution));
+			assert_eq!(balances(&99), (95, 5));
+
+			// no good solution was stored.
+			assert!(!MultiPhase::finalize_signed_phase().0);
+			// and the bond is gone.
+			assert_eq!(balances(&99), (95, 0));
+		})
+	}
+
+	#[test]
+	fn suppressed_solution_gets_bond_back() {
+		ExtBuilder::default().build_and_execute(|| {
+			roll_to(15);
+			assert!(MultiPhase::current_phase().is_signed());
+
+			let mut solution = raw_solution();
+			assert_eq!(balances(&99), (100, 0));
+			assert_eq!(balances(&999), (100, 0));
+
+			// submit as correct.
+			assert_ok!(submit_with_witness(Origin::signed(99), solution.clone()));
+
+			// make the solution invalid and weaker.
+			solution.score[0] -= 1;
+			assert_ok!(submit_with_witness(Origin::signed(999), solution));
+			assert_eq!(balances(&99), (95, 5));
+			assert_eq!(balances(&999), (95, 5));
+
+			// _some_ good solution was stored.
+			assert!(MultiPhase::finalize_signed_phase().0);
+
+			// 99 is rewarded.
+			assert_eq!(balances(&99), (100 + 7, 0));
+			// 999 gets everything back.
+			assert_eq!(balances(&999), (100, 0));
+		})
+	}
+
+	#[test]
+	fn cannot_submit_worse_with_full_queue() {
+		ExtBuilder::default().build_and_execute(|| {
+			roll_to(15);
+			assert!(MultiPhase::current_phase().is_signed());
+
+			for s in 0..SignedMaxSubmissions::get() {
+				// score is always getting better
+				let solution = RawSolution { score: [(5 + s).into(), 0, 0], ..Default::default() };
+				assert_ok!(submit_with_witness(Origin::signed(99), solution));
+			}
+
+			// weaker.
+			let solution = RawSolution { score: [4, 0, 0], ..Default::default() };
+
+			assert_noop!(
+				submit_with_witness(Origin::signed(99), solution),
+				Error::<Runtime>::SignedQueueFull,
+			);
+		})
+	}
+
+	#[test]
+	fn weakest_is_removed_if_better_provided() {
+		ExtBuilder::default().build_and_execute(|| {
+			roll_to(15);
+			assert!(MultiPhase::current_phase().is_signed());
+
+			for s in 0..SignedMaxSubmissions::get() {
+				// score is always getting better
+				let solution = RawSolution { score: [(5 + s).into(), 0, 0], ..Default::default() };
+				assert_ok!(submit_with_witness(Origin::signed(99), solution));
+			}
+
+			assert_eq!(
+				MultiPhase::signed_submissions()
+					.into_iter()
+					.map(|s| s.solution.score[0])
+					.collect::<Vec<_>>(),
+				vec![9, 8, 7, 6, 5]
```

Comment:
I wonder: would it not be better to store this reverse? because at teh end we want to `.pop()` from this function, and we want the best to be last.

Reply:
As I understand it, it's better to store it in the current order, which is reverse-sorted. Rationale: for every `insert_submission` except the first `T::MaxSignedSubmissions`, we need to remove the lowest, which is `O(1)` in this order and `O(n)` if we stored things forward-sorted.

It might, however, be worth considering storing the submission queue as a `BoundedBTreeSet` instead of a `Vec`. We'd get `O(log n)` insertions, deletions, and access to the ends. Depending on the precise workload, I'd expect its worst case to be more performant than a vector for this application.

For that matter, given that we already keep track of `MaxSignedSubmissions`, one or another of the bounded data structures seems well-worth using, instead of handling overflow logic manually here.

Reply:
This idea is taking too much time time to implement, so I'm moving that to a successor issue: https://github.com/paritytech/substrate/pull/8933.

Reply:
yeah, probably not a big deal, the queue can have a limited (in the order of dozens) of slots, in which all possibilities are almost identical.